backend/src/services/BaseService.ts

```diff
@@ -0,0 +1,117 @@
+import { loggers, constants, resolveSchema } from '@projecttacoma/node-fhir-server-core';
+import { BadRequestError } from '../util/errorUtils';
+import { checkContentTypeHeader } from '../util/inputUtils';
+import { v4 as uuidv4 } from 'uuid';
+import { createResource, updateResource } from '../db/dbOperations';
+import path from 'path';
+import { DetailedEntry, replaceReferences } from '../util/baseUtils';
+
+const logger = loggers.get('default');
+
+/**
+ * Uploads a transaction bundle to the server and returns the transaction-response bundle.
+ */
+export async function uploadTransactionBundle(req: any, res: any) {
+  logger.info('POST /');
+  logger.info(`Base >>> transaction`);
+  const contentType: string | undefined = req.headers['content-type'];
+  checkContentTypeHeader(contentType);
+
+  const entries: DetailedEntry[] = req.body.entry;
+  const baseVersion: keyof typeof constants.VERSIONS = req.params.base_version;
+
+  if (req.body.resourceType !== 'Bundle') {
+    throw new BadRequestError(`Expected 'resourceType: Bundle'. Received 'resourceType: ${req.body.resourceType}'.`);
+  }
+
+  if (req.body.type.toLowerCase() != 'transaction') {
+    throw new BadRequestError(`Expected 'type: transaction. Received type: ${req.body.type}'.`);
+  }
+
+  const requestResults = await uploadResourcesFromBundle(entries);
+
+  const bundle = makeTransactionResponseBundle(requestResults, res, baseVersion);
+  logger.info('Transaction bundle successfully uploaded.');
+  return bundle;
+}
+
+/**
+ * Uploads Library and Measure resources from a transaction bundle to the server.
+ */
+async function uploadResourcesFromBundle(entries: DetailedEntry[]) {
+  logger.info(`Inserting Measure and Library resources from transaction bundle`);
+  const scrubbedEntries = replaceReferences(entries);
+
+  const requestsArray = scrubbedEntries.map(async entry => {
+    const { method } = entry.request as fhir4.BundleEntryRequest;
```

Comment:
I also wasn't totally sure about this- I thought this was okay because as long as an error is thrown when `entry.request` is undefined then it aligns with the spec...but if you think throwing a BadRequestError here instead would be more helpful for the user then I can add that!